### PR TITLE
Split kill_house() fix issues #1234

### PR DESCRIPTION
**Pull request by wevsty**
Details View issues #1234

### CODE REVIEW

**Comment by selwin** [comment]
Can you try to come up with a test to verify this behavior? Changing this area of the code is always a bit risky so I want to be extra careful to see that exit codes are handled correctly.

Also, we'll need all tests to pass before we merge this in.

**Comment by wevsty** [comment]
> 
> 
> Can you try to come up with a test to verify this behavior? Changing this area of the code is always a bit risky so I want to be extra careful to see that exit codes are handled correctly.
> 
> Also, we'll need all tests to pass before we merge this in.

Only one part of the test failed.
According to the logs, it is in HerokuWorkerShutdownTestCase.test_ handle_shutdown_request in this function.
`
__________ HerokuWorkerShutdownTestCase.test_handle_shutdown_request ___________

self = <tests.test_worker.HerokuWorkerShutdownTestCase testMethod=test_handle_shutdown_request>
mock_logger_info = <MagicMock name='info' id='139955829528224'>

    @mock.patch('rq.worker.logger.info')
    def test_handle_shutdown_request(self, mock_logger_info):
        """Mutate HerokuWorker so _horse_pid refers to an artificial process
        and test handle_warm_shutdown_request"""
        w = HerokuWorker('foo')
    
        path = os.path.join(self.sandbox, 'shouldnt_exist')
        p = Process(target=create_file_after_timeout, args=(path, 2))
        p.start()
        self.assertEqual(p.exitcode, None)
    
        w._horse_pid = p.pid
        w.handle_warm_shutdown_request()
        p.join(2)
        # would expect p.exitcode to be -34 but for some reason os.waitpid is setting it to None, even though
        # the process has ended
>       self.assertEqual(p.exitcode, None)
E       AssertionError: -34 != None

tests/test_worker.py:1217: AssertionError
`
In particular, the comment states that the test was written with the expected value of -34, but that the test returned a value of None.
The change resulted in what the developer expected, so I think the change itself should be fine.
Maybe the failing test issue requires a change in the test script, but I'd like to hear your thoughts before doing so.

**Comment by selwin** [comment]
I think the current approach (splitting the kill and wait) makes a ton of sense. Catching `ChildProcessError` is also 👍 since if the process is already gone, it's safe for the worker to continue.

If possible, I'd like a test case to cover issues raised in https://github.com/rq/rq/issues/1234 .

> In particular, the comment states that the test was written with the expected value of -34, but that the test returned a value of None.
The change resulted in what the developer expected, so I think the change itself should be fine.

You're right. So this is actually a good change.

File: rq/worker.py [hunk 1]
```diff
@@ -378,15 +378,27 @@ def kill_horse(self, sig=SIGKILL):
...
-            os.waitpid(self.horse_pid, 0)
-            self.log.info('Killed horse pid %s', self.horse_pid)
         except OSError as e:
             if e.errno == errno.ESRCH:
                 # "No such process" is fine with us
                 self.log.debug('Horse already dead')
             else:
                 raise
 
+    def wait_horse(self):
```
Comment (selwin) [inline rq/worker.py:388]: Can we rename this to `wait_for_horse()`?

**Comment by wevsty** [comment]
> 
> 
> I think the current approach (splitting the kill and wait) makes a ton of sense. Catching `ChildProcessError` is also 👍 since if the process is already gone, it's safe for the worker to continue.
> 
> If possible, I'd like a test case to cover issues raised in #1234 .
> 
> > In particular, the comment states that the test was written with the expected value of -34, but that the test returned a value of None.
> > The change resulted in what the developer expected, so I think the change itself should be fine.
> 
> You're right. So this is actually a good change.

I've made the changes visually and the latest commit has passed the test.

But I don't have a good idea about adding new tests to validate this.
I think the original test could have covered this issue, it was just ignored before.

File: rq/worker.py [hunk 2]
```diff
@@ -387,6 +386,20 @@ def kill_horse(self, sig=SIGKILL):
             else:
                 raise
 
+    def wait_for_horse(self):
+        """
+        A waiting the end of the horse process and recycling resources.
+        """
+        try:
+            pid = None
+            stat = None
```
Comment (selwin) [inline rq/worker.py:395]: `pid` and `stat` should be declared outside of the `try/except` block.

File: rq/worker.py [hunk 3]
```diff
@@ -387,6 +386,20 @@ def kill_horse(self, sig=SIGKILL):
...
+            if self.horse_pid:
```
Comment (selwin) [inline rq/worker.py:396]: No need to check for the existence of `horse_pid` since this method is only called by the main worker.

Comment (wevsty) [reply rq/worker.py]: These codes have been adjusted.

**Pull request merged**